File: Java/refactoring_exercise_2/src/test/java/za/co/entlect/refactoring2/controller/StrategyExample.java
package za.co.entlect.refactoring2.controller;

import java.util.HashMap;
import java.util.Map;

public class StrategyExample {

    public static final String ADD = "ADD";
    public static final String SUBTRACT = "SUBTRACT";
    public static final String MULTIPLY = "MULTIPLY";
    public static final String DIVIDE = "DIVIDE";

    private Map<String, NumericStrategy> numericStrategyMap = new HashMap<String, NumericStrategy>();

    public StrategyExample(){
        numericStrategyMap.put(ADD, new AddStrategy());
        numericStrategyMap.put(SUBTRACT, new SubtractStrategy());
        numericStrategyMap.put(MULTIPLY, new MultiplyStrategy());
        numericStrategyMap.put(DIVIDE, new DivideStrategy());
    }

    public int calculate(String operation, int a, int b){
        return numericStrategyMap.get(operation).calculate(a, b);
    }

    interface NumericStrategy {
        int calculate(int a, int b);
    }

    class AddStrategy implements NumericStrategy {

        @Override
        public int calculate(int a, int b) {
            return a+b;
        }
    }

    class SubtractStrategy implements NumericStrategy {

        @Override
        public int calculate(int a, int b) {
            return a-b;
        }
    }

    class MultiplyStrategy implements NumericStrategy {

        @Override
        public int calculate(int a, int b) {
            return a*b;
        }
    }

    class DivideStrategy implements NumericStrategy{

        @Override
        public int calculate(int a, int b) {
            return a/b;
        }
    }

}

File: C#/Refactoring/refactoring_exercise_final/test/za/co/entelect/refactoring_final/controller/StrategyExample.cs
using System;
using System.Collections.Generic;

namespace refactoring_exercise_final.test.za.co.entelect.refactoring_final.controller
{
    public class StrategyExample
    {

        public const string Add = "ADD";
        public const string Subtract = "SUBTRACT";
        public const string Multiply = "MULTIPLY";
        public const string Divide = "DIVIDE";

        private Dictionary<String, INumericStrategy> numericStrategyMap = new Dictionary<String, INumericStrategy>();

        public StrategyExample()
        {
            numericStrategyMap.Add(Add, new AddStrategy());
            numericStrategyMap.Add(Subtract, new SubtractStrategy());
            numericStrategyMap.Add(Multiply, new MultiplyStrategy());
            numericStrategyMap.Add(Divide, new DivideStrategy());
        }

        public int Calculate(String operation, int a, int b)
        {
            return numericStrategyMap[operation].Calculate(a, b);
        }

        private interface INumericStrategy
        {
            int Calculate(int a, int b);
        }

        private class AddStrategy : INumericStrategy
        {
            public int Calculate(int a, int b)
            {
                return a + b;
            }
        }

        private class SubtractStrategy : INumericStrategy
        {

            public int Calculate(int a, int b)
            {
                return a - b;
            }
        }

        private class MultiplyStrategy : INumericStrategy
        {
            public int Calculate(int a, int b)
            {
                return a*b;
            }
        }

        private class DivideStrategy : INumericStrategy
        {

            public int Calculate(int a, int b)
            {
                return a/b;
            }
        }

    }
}